Update dynamic instrumentation docs - log probes (#16905)
DataDog/documentation: content/en/dynamic_instrumentation/_index.md

- [Datadog Agent][1] 7.41.1 or higher is installed alongside your service.
- [Remote Configuration][2] is enabled in that Agent.

> - For Java applications, tracing library [`dd-trace-java`][3] 1.8.0 or higher.

Author: I don't understand what you mean 🙂 I pushed another commit yesterday with version updates.

All versions are set to latest, I think that's fine since they include all recent bugfixes. We will need to update the python version once we release all the updates we are working on now.

Dynamic Instrumentation can help you understand what your application is doing at runtime. By adding a Dynamic Instrumentation probe you are exporting additional data from your application, without the need to do any code change or redeployment.

### Creating a probe

Both log and metric probes require the same initial setup:

1. Go to the [Dynamic Instrumentation page][12].
1. Click **Create Probe** in the top right, or click the three dot context menu on a service and select **Add a probe for this service**.
1. If not prefilled, choose a service from the list.
1. If not prefilled, choose runtime, environment and version.
1. In the source code, select a class and method or a source file and line where to set the probe.

>   1. If you set up Source Code Integration for your service, when setting line probes you will receive autocomplete suggestions for the source file and you will see the source code of the selected file when choosing the line.

Reviewer: I think we better mention here that in order to utilize the Source Code Integration the user has to select specific env and version (rather than "All" option) otherwise we cannot get the matching commit sha, i.e. no code display.

Author: We talked about this, we're actually changing the UI to automatically pick the latest version to display source code from - even if the user selects 'all', they will still see something.

I think we show an error in the UI anyway if we don't have enough info to fetch source code, so I think we can leave this as is 🙂

> 1. In the source code, select a class and method or a source file and line where to set the probe.
>   1. If you set up Source Code Integration for your service, when setting line probes you will receive autocomplete suggestions for the source file and you will see the source code of the selected file when choosing the line.

Reviewer 2: ```suggestion
1. In the source code, specify where to set the probe by selecting either a class and method or a source file and line. 
   If you set up Source Code Integration for your service, autocomplete shows suggestions for the selecting a file and displays the file's code so you can choose the line.
```

### Creating a log probe

> A *log probe* will emit a log when it executes.

Reviewer 2: ```suggestion
A *log probe* emits a log when it executes.
```

> Log probes without snapshot capture are enabled on all service instances that match the desired environment and version. They are rate limited to 5000 logs/s.

Author: :+1: Good call, added this.

> If you enable `Capture method parameters and local variables` on the log probe, it will also capture method arguments, local variables, class fields, the call stack and any exceptions from the execution context and add them to the log event - you can see the captured values in the Datadog UI. Since capturing this data is a more performance intensive operation, it is only enabled on one instance of your service which matches the desired environment and version configured on the probe. Probes with capture are rate limited to execute once every second.

Reviewer 2: ```suggestion
If you enable `Capture method parameters and local variables` on the log probe, it also captures the following values from the execution context and adds them to the log event:
- method arguments
- local variables
- class fields
- the call stack
- exceptions 
You can see the captured values in the Datadog UI. 

Because capturing this data is performance-intensive, it is enabled on only one instance of your service that matches the probe's environment and version settings. Probes with capture enabled are rate limited to execute once per second.
```

> If you enable `Capture method parameters and local variables` on the log probe, it will also capture method arguments, local variables, class fields, the call stack and any exceptions from the execution context and add them to the log event - you can see the captured values in the Datadog UI. Since capturing snapshots is a more performance intensive operation, by default it is only enabled on one instance of your service that matches the desired environment and version. Probes with snapshot capture are rate limited to execute once every second.

Author: I wouldn't add another list here, we already have three on the page 🙂 I think having two sentences describing this is fine.

> Log probes without extra data capturing are enabled on all service instances that match the desired environment and version. They are rate limited to execute at most 5000 times per second, on each service instance.

Reviewer 2: ```suggestion
Log probes without extra data capturing are enabled on all service instances that match the specified environment and version. They are rate limited to execute at most 5000 times per second, on each service instance.
```

For more information, read [How Dynamic Instrumentation Works][11].

To create a log probe:

> 1. Complete the generic probe setup (choose service, environment, version and probe location).

Reviewer 2: ```suggestion
1. Complete the generic probe setup (choose service, environment, version, and probe location).
```

1. Select **Log** as the probe type.
1. Define a log message template. You can use the Dynamic Instrumentation expression language to reference values from the execution context.

> 1. Select if you would like to enable extra data capturing from the probe.

Reviewer 2: ```suggestion
1. Optionally enable extra data capturing from the probe.
```

> 1. (optional) Define a condition using the Dynamic Instrumentation expression language. The log will only be emitted when the expression evaluates to true.

Reviewer 2: ```suggestion
1. Optionally define a condition using the Dynamic Instrumentation expression language. The log is emitted when the expression evaluates to true.
```

{{< img src="dynamic_instrumentation/log_probe.png" alt="Creating a Dynamic Instrumentation log probe" >}}

### Creating a metric probe

> A *metric probe* will emit a metric when it executes.

Reviewer 2: ```suggestion
A *metric probe* emits a metric when it executes.
```

Metric probes are automatically enabled on all service instances that match the configured environment and version.

> You can use the Dynamic Instrumentation expression language to reference numeric values from the context, such as a variable, class field or an expression that yields a numeric value.

Reviewer 2: ```suggestion
You can use the Dynamic Instrumentation expression language to reference numeric values from the context, such as a variable, a class field, or an expression that yields a numeric value.
```

For more information, read [How Dynamic Instrumentation Works][11].

To create a metric probe:

> 1. Complete the generic probe setup (choose service, environment, version and probe location).

Reviewer 2: ```suggestion
1. Complete the generic probe setup (choose service, environment, version, and probe location).
```

1. Select **Metric** as the probe type.

> 1. Specify a name for the metric. The metric name will be prefixed with `dynamic.instrumentation.metric.probe.`.

Reviewer 2: ```suggestion
1. Specify a name for the metric, which will be prefixed with `dynamic.instrumentation.metric.probe.`.
```

> 1. Select a metric type (count, gauge or histogram).

Reviewer 2: ```suggestion
1. Select a metric type (count, gauge, or histogram).
```

> 1. Choose the value of the metric using the Debugger expression language. For count metrics this is optional, if omitted every invocation will increase the count by one.

Reviewer 2: ```suggestion
1. Choose the value of the metric using the Debugger expression language. For count metrics this is optional, and if you omit it, every invocation increments the count by one.
```
